File: android/app/src/main/java/run/xrun/xrunapp/AdisonModule.kt
package run.xrun.xrunapp

import android.util.Log
import co.adison.offerwall.Adison
import co.adison.offerwall.AdisonConfig
import co.adison.offerwall.AdisonListType
import co.adison.offerwall.AdisonThemeMode
import co.adison.offerwall.Gender
import co.adison.offerwall.Server
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class AdisonModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var isInitialized = false

    override fun getName(): String = "AdisonModule"

    @ReactMethod
    fun initialize(appKey: String, server: String, promise: Promise) {
        try {
            if (isInitialized) {
                promise.resolve(true)
                return
            }
            val context = reactApplicationContext.applicationContext
            Adison.initialize(context, appKey)
            Adison.setServer(
                if (server.equals("development", ignoreCase = true)) Server.DEVELOPMENT
                else Server.PRODUCTION
            )
            isInitialized = true
            Log.d("AdisonModule", "initialize ok: server=$server")
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e("AdisonModule", "initialize failed", e)
            promise.reject("INIT_ERROR", e.message ?: "Unknown error")
        }
    }

    @ReactMethod
    fun setConfig(
        offerwallTitle: String,
        listType: String, 
        themeMode: String 
    ) {
        try {
            val config = AdisonConfig().apply {
                this.offerwallListTitle = offerwallTitle
                this.listType = if (listType.equals("FEED", ignoreCase = true))
                    AdisonListType.FEED else AdisonListType.LIST
                this.prepareViewHidden = true
                this.themeMode = when (themeMode) {
                    "Dark" -> AdisonThemeMode.Dark
                    "System" -> AdisonThemeMode.System
                    else -> AdisonThemeMode.Light
                }
            }
            Adison.setConfig(config)
            Log.d("AdisonModule", "setConfig ok")
        } catch (e: Exception) {
            Log.e("AdisonModule", "setConfig failed", e)
        }
    }

    @ReactMethod
    fun setUid(uid: String) {
        try {
            Adison.setUid(uid)
            Log.d("AdisonModule", "setUid ok")
        } catch (e: Exception) {
            Log.e("AdisonModule", "setUid failed", e)
        }
    }

    @ReactMethod
    fun unsetUid() {
        try {
            Adison.setUid(null)
            Log.d("AdisonModule", "unsetUid ok")
        } catch (e: Exception) {
            Log.e("AdisonModule", "unsetUid failed", e)
        }
    }

    @ReactMethod
    fun setTargeting(birthYear: Int, gender: String?) {
        try {
            if (birthYear > 0) {
                Adison.setBirthYear(birthYear)
            }
            when (gender?.uppercase()) {
                "M" -> Adison.setGender(Gender.MALE)
                "F" -> Adison.setGender(Gender.FEMALE)
                else -> Adison.setGender(Gender.UNKNOWN)
            }
        } catch (e: Exception) {
            Log.w("AdisonModule", "setTargeting failed: ${e.message}")
        }
    }

    @ReactMethod
    fun availableReward(promise: Promise) {
        try {
            Adison.availableReward { name, unit, points ->
                val result = com.facebook.react.bridge.Arguments.createMap().apply {
                    putString("name", name)
                    putString("unit", unit)
                    putInt("points", points)
                }
                promise.resolve(result)
            }
        } catch (e: Exception) {
            Log.e("AdisonModule", "availableReward failed", e)
            promise.reject("REWARD_ERROR", e.message ?: "Unknown error")
        }
    }

    @ReactMethod
    fun showOfferwall(promise: Promise) {
        try {
            if (!isInitialized) {
                promise.reject("NOT_INITIALIZED", "Adison SDK not initialized")
                return
            }
            Adison.showOfferwall()
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e("AdisonModule", "showOfferwall failed", e)
            promise.reject("SHOW_ERROR", e.message ?: "Unknown error")
        }
    }

    @ReactMethod
    fun showOfferwallAd(adId: Int, keepParent: Boolean, promise: Promise) {
        try {
            if (!isInitialized) {
                promise.reject("NOT_INITIALIZED", "Adison SDK not initialized")
                return
            }
            Adison.showOfferwall(adId, keepParent)
            promise.resolve(true)
        } catch (e: Exception) {
            Log.e("AdisonModule", "showOfferwallAd failed", e)
            promise.reject("SHOW_ERROR", e.message ?: "Unknown error")
        }
    }
}

File: ios/XRUN/AdisonModule.swift
import Foundation
import React
import UIKit
import AdisonOfferwallSDK

@objc(AdisonModule)
class AdisonModule: NSObject {

  private var isInitialized = false

  @objc
  static func requiresMainQueueSetup() -> Bool { return true }

  @objc
  func initialize(
    _ appKey: String,
    server: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      if self.isInitialized {
        resolve(true)
        return
      }

      Adison.shared.initialize(appId: appKey)
      if server.lowercased() == "development" {
        Adison.shared.setServer(.development)
      } else {
        Adison.shared.setServer(.production)
      }
      self.isInitialized = true
      NSLog("[Adison] initialize ok: server=%@", server)
      resolve(true)
    }
  }

  @objc
  func setConfig(
    _ offerwallTitle: String,
    listType: String,
    themeMode: String
  ) {
    DispatchQueue.main.async {
      let config = AdisonConfig()
      config.offerwallListTitle = offerwallTitle
      config.listType = (listType.uppercased() == "FEED") ? .feed : .list
      config.prepareViewHidden = true
      switch themeMode {
      case "Dark":   config.themeMode = .dark
      case "System": config.themeMode = .system
      default:       config.themeMode = .light
      }
      Adison.shared.config = config
    }
  }

  @objc
  func setUid(_ uid: String) {
    DispatchQueue.main.async {
      Adison.shared.uid = uid
    }
  }

  @objc
  func unsetUid() {
    DispatchQueue.main.async {
      Adison.shared.uid = nil
    }
  }

  @objc
  func setTargeting(_ birthYear: NSNumber, gender: NSString?) {
    DispatchQueue.main.async {
      if birthYear.intValue > 0 {
        Adison.shared.birthYear = birthYear.intValue
      }
      if let g = gender as String? {
        switch g.uppercased() {
        case "M": Adison.shared.gender = .male
        case "F": Adison.shared.gender = .female
        default: break
        }
      }
    }
  }

  @objc
  func showOfferwall(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      guard self.isInitialized else {
        reject("NOT_INITIALIZED", "Adison SDK not initialized", nil)
        return
      }
      guard let rootVC = UIApplication.shared.keyWindow?.rootViewController else {
        reject("NO_ROOT_VC", "No root view controller", nil)
        return
      }
      Adison.shared.presentOfferwall(from: rootVC)
      resolve(true)
    }
  }

  @objc
  func showOfferwallAd(
    _ adId: NSNumber,
    keepParent: Bool,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      guard self.isInitialized else {
        reject("NOT_INITIALIZED", "Adison SDK not initialized", nil)
        return
      }
      guard let rootVC = UIApplication.shared.keyWindow?.rootViewController else {
        reject("NO_ROOT_VC", "No root view controller", nil)
        return
      }
      Adison.shared.presentOfferwall(from: rootVC, adId: adId.intValue, keepParent: keepParent)
      resolve(true)
    }
  }

  @objc
  func availableReward(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.main.async {
      Adison.shared.availableReward { name, unit, points in
        resolve([
          "name": name ?? "",
          "unit": unit ?? "",
          "points": points,
        ])
      }
    }
  }
}
